sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParseDriver.scala

```diff
@@ -127,7 +127,7 @@ abstract class AbstractSqlParser extends ParserInterface with SQLConfHelper with
         toResult(parser)
       }
       catch {
-        case e: ParseCancellationException =>
+        case _: ParseCancellationException | _: ParseException =>
```

Comment:
This fix looks too broad.

Have done some investigations, the key point should be to use `BailErrorStrategy`.

This should be a long-standing issue, before SPARK-38385, Spark uses `DefaultErrorStrategy`, and after SPARK-38385 Spark uses `class SparkParserErrorStrategy() extends DefaultErrorStrategy`.

So this is not the correct implementation of the "two-stage parsing strategy" 

https://github.com/antlr/antlr4/issues/192#issuecomment-15238595

Reply:
Thanks for remind. The `SparkParserErrorStrategy` will rewrap Exception which throw by antlr4. And finally throw it by `ParseException`. So if we want to fix this, catch `ParserException` will be a way. If you have any better idea please let me know. Thanks!

Reply:
@Hisoka-X @cloud-fan I think we should provide two `ErrorStrategy`s
```
class SparkParserBailErrorStrategy() extends BailErrorStrategy
class SparkParserDefaultErrorStrategy() extends DefaultErrorStrategy
```

and following the https://github.com/antlr/antlr4/issues/192#issuecomment-15238595 to implement the "two-stage parsing strategy"

Reply:
> @Hisoka-X @cloud-fan I think we should provide two `ErrorStrategy`s
> 
> ```
> class SparkParserBailErrorStrategy() extends BailErrorStrategy
> class SparkParserDefaultErrorStrategy() extends DefaultErrorStrategy
> ```
> 
> and follow the [antlr/antlr4#192 (comment)](https://github.com/antlr/antlr4/issues/192#issuecomment-15238595) to implement the "two-stage parsing strategy"

I'm not sure this way will be worked. The main reason for custom `ErrorStrategy` are replace anltr4 Exception by some Spark Exception. So we need catch the Spark Exception too.

Reply:
@pan3793 can you make a PR to implement your idea and see if it works?